packages/oats-runtime/src/make.ts

```diff
@@ -533,6 +445,113 @@ export function makeNullable(maker: any) {
   };
 }
 
+function isScalar(type: Type): boolean {
+  if (type.type === 'named') {
+    return isScalar(type.reference.definition);
+  }
+  return ['array', 'object', 'union', 'intersection', 'unknown'].indexOf(type.type) < -1;
+}
+
+function isEnum(type: Type) {
+  if (type.type === 'null') {
+    return true;
+  }
+  if (
+    type.type === 'string' ||
+    type.type === 'boolean' ||
+    type.type === 'number' ||
+    type.type === 'integer'
+  ) {
+    return !!type.enum;
+  }
+  return false;
+}
+
+function priority(v: ObjectType['properties'][0]) {
+  // check scalars first to avoid constructing trees unnecessarily
+  if (isScalar(v.value)) {
+    if (isEnum(v.value)) {
+      // high chance this is a union type tag
+      if ((v.value as any).enum?.length === 1) {
+        return 4;
+      }
+      // enum matches less things than non enum
+      return 3;
+    }
+    if (v.value.type === 'string') {
+      // a pattern matches less things than a non pattern so lets try this first
+      if (v.value.pattern) {
+        return 3;
+      }
+    }
+    if (v.required) {
+      return 2;
+    }
+    return 1;
+  }
+  return 0;
+}
+
+function fromObjectReflection(type: ObjectType): Maker<any, any> {
+  const comparisonOrder = Object.keys(type.properties).sort((aKey, bKey) => {
+    const a = type.properties[aKey]!;
+    const b = type.properties[bKey]!;
+    return priority(b) - priority(a);
```

Comment:
higher priority gets tried first

Reply:
a bit obscure that we use inverse priority here. easy to miss that we're sorting asc instead of desc
might be clearer if `0` was top priority, but this might be one of those topics like how to pronounce `gif` 🌝

Reply:
de-invert'd